### PR TITLE
Add event recorder for reconciler

### PR DESCRIPTION
**Pull request by zhangtbj**
1, Add Event recorder for Reconciler to record fail at BOSHDeployment CR level
2, Add an integration test for the event message test
3, All test passed:
```
Ran 9 of 9 Specs in 3.179 seconds
SUCCESS! -- 9 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

### CODE REVIEW

**Comment by zhangtbj** [comment]
We will also provide a doc about how to use event recorder and a common solution for all components by using another story:
https://www.pivotaltracker.com/n/projects/2192232

File: integration/deploy_test.go [hunk 1]
```diff
@@ -55,6 +55,28 @@ var _ = Describe("Deploy", func() {
 	})
 
 	Context("when incorrectly setup", func() {
+		It("should report correct error if ops for is incorrect", func() {
```
Comment (manno) [inline integration/deploy_test.go]: I don't understand the description, "ops for is incorrect". Ops for what :) ?

File: integration/deploy_test.go [hunk 2]
```diff
@@ -55,6 +55,28 @@ var _ = Describe("Deploy", func() {
...
+			Expect(err).NotTo(HaveOccurred())
+			defer tearDown()
+
+			boshDeployment, tearDown, err := env.CreateBOSHDeployment(env.Namespace, env.InterpolateBOSHDeployment("test", "manifest", "bosh-ops", "bosh-ops-secret"))
+			Expect(err).NotTo(HaveOccurred())
+			defer tearDown()
+
+			success, result := env.GetBOSHDeploymentEventMessage(env.Namespace, boshDeployment.ObjectMeta.Name, string(boshDeployment.ObjectMeta.UID))
+			Expect(success).To(BeTrue())
+			Expect(result).To(ContainSubstring("Expected to find exactly one matching array item for path '/instance-groups/name=api' but found 0"))
```
Comment (manno) [inline integration/deploy_test.go]: Does this test one of the events from `reconciler.go`? Which one?

File: pkg/kube/controllers/boshdeployment/reconciler.go [hunk 3]
```diff
@@ -66,25 +69,30 @@ func (r *ReconcileBOSHDeployment) Reconcile(request reconcile.Request) (reconcil
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
+		r.recorder.Event(instance, corev1.EventTypeWarning, "GetCRD Error", err.Error())
```
Comment (manno) [inline pkg/kube/controllers/boshdeployment/reconciler.go]: This one is missing a test?

File: pkg/kube/controllers/boshdeployment/reconciler.go [hunk 4]
```diff
@@ -66,25 +69,30 @@ func (r *ReconcileBOSHDeployment) Reconcile(request reconcile.Request) (reconcil
...
 		return reconcile.Result{}, err
 	}
 
 	// retrieve manifest
 	manifest, err := r.resolver.ResolveCRD(instance.Spec, request.Namespace)
 	if err != nil {
+		r.recorder.Event(instance, corev1.EventTypeWarning, "ResolveCRD Error", err.Error())
```
Comment (manno) [inline pkg/kube/controllers/boshdeployment/reconciler.go]: I think this is tested

File: pkg/kube/controllers/boshdeployment/reconciler.go [hunk 5]
```diff
@@ -66,25 +69,30 @@ func (r *ReconcileBOSHDeployment) Reconcile(request reconcile.Request) (reconcil
...
 		return reconcile.Result{}, err
 	}
 
 	// TODO validation
 	if len(manifest.InstanceGroups) < 1 {
-		return reconcile.Result{}, fmt.Errorf("manifest is missing instance groups")
+		err = fmt.Errorf("manifest is missing instance groups")
+		r.recorder.Event(instance, corev1.EventTypeWarning, "MissingInstance Error", err.Error())
```
Comment (manno) [inline pkg/kube/controllers/boshdeployment/reconciler.go]: This should have a test. We can also remove the TODO comment.

File: pkg/kube/controllers/boshdeployment/reconciler.go [hunk 6]
```diff
@@ -66,25 +69,30 @@ func (r *ReconcileBOSHDeployment) Reconcile(request reconcile.Request) (reconcil
...
+		return reconcile.Result{}, err
 	}
 
 	// Define a new Pod object
 	pod := newPodForCR(manifest, request.Namespace)
 
 	// Set BOSHDeployment instance as the owner and controller
 	if err := r.setReference(instance, pod, r.scheme); err != nil {
+		r.recorder.Event(instance, corev1.EventTypeWarning, "NewPodForCR Error", err.Error())
```
Comment (manno) [inline pkg/kube/controllers/boshdeployment/reconciler.go]: I don't think any of the 'new pod' events needs a test. The pod functionality is a left over and will probably be removed in the future. 
The BOSHDeployment controller will serve as an entry point, to create the other resources (extended*). But I don't know which story that belongs to and I'm hesitant to remove the newPodCR code and existing tests.

File: integration/deploy_test.go (hunk 2 above)
Comment (edwardstudy) [reply integration/deploy_test.go]: It tested this event: https://github.com/cloudfoundry-incubator/cf-operator/blob/f5abafcd7742f3eeecef3b4305279f1140ce18b6/pkg/kube/controllers/boshdeployment/reconciler.go#L79

I'll update other events

File: pkg/kube/controllers/boshdeployment/reconciler.go (hunk 6 above)
Comment (edwardstudy) [reply pkg/kube/controllers/boshdeployment/reconciler.go]: I think so.

File: pkg/kube/controllers/boshdeployment/reconciler.go (hunk 4 above)
Comment (edwardstudy) [reply pkg/kube/controllers/boshdeployment/reconciler.go]: Updated test cases

File: pkg/kube/controllers/boshdeployment/reconciler.go (hunk 3 above)
Comment (edwardstudy) [reply pkg/kube/controllers/boshdeployment/reconciler.go]: Updated test cases

**Pull request merged**